File: scripts/discord-control-bot.mjs
#!/usr/bin/env node
import { readFileSync, existsSync } from "node:fs";
import process from "node:process";

function loadEnvFile(path) {
  if (!existsSync(path)) return;
  const raw = readFileSync(path, "utf8");
  for (const line of raw.split(/\r?\n/)) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith("#")) continue;
    const eq = trimmed.indexOf("=");
    if (eq < 0) continue;
    const key = trimmed.slice(0, eq).trim();
    if (!key || process.env[key] !== undefined) continue;
    let value = trimmed.slice(eq + 1).trim();
    if (
      (value.startsWith('"') && value.endsWith('"')) ||
      (value.startsWith("'") && value.endsWith("'"))
    ) {
      value = value.slice(1, -1);
    }
    process.env[key] = value;
  }
}

loadEnvFile(".env.local");

const config = {
  token: process.env.DISCORD_BOT_TOKEN,
  guildId: process.env.DISCORD_GUILD_ID,
  allowedChannelNames: ["filings", "filings2"],
  filingsChannelId: process.env.DISCORD_FILINGS_CHANNEL_ID,
  filings2ChannelId: process.env.DISCORD_FILINGS2_CHANNEL_ID,
  apiBaseUrl: process.env.BEACON_API_BASE_URL || "http://127.0.0.1:3000",
  pollMs: Number(process.env.DISCORD_BOT_POLL_MS || 4000),
  debugEnabled: (process.env.DISCORD_BOT_DEBUG || "false") === "true"
};

if (!config.token) throw new Error("Missing DISCORD_BOT_TOKEN in .env.local");
if (!config.guildId) throw new Error("Missing DISCORD_GUILD_ID in .env.local");

const discordApi = "https://discord.com/api/v10";
const authHeaders = {
  Authorization: `Bot ${config.token}`,
  "Content-Type": "application/json"
};

async function sendDiscord(content) {
  const url = process.env.DISCORD_WEBHOOK_URL;
  if (!url) throw new Error("Missing DISCORD_WEBHOOK_URL");

  const payload = { content };
  const res = await fetch(url, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload)
  });

  if (!res.ok) {
    const txt = await res.text().catch(() => "");
    throw new Error(`Discord webhook failed: ${res.status} ${txt}`);
  }
}

const trackedChannelIds = [config.filingsChannelId, config.filings2ChannelId].filter(Boolean);
const lastSeenMessageId = new Map();

// ---- state tracking (NEW) ----
let pollCount = 0;     // counts "poll cycles" (one setInterval tick)
let botId = "<unknown>";
let isTickRunning = false;
// ------------------------------

function maskToken(token) {
  if (!token) return "<missing>";
  if (token.length <= 12) return `${token.slice(0, 2)}***`;
  return `${token.slice(0, 6)}...${token.slice(-4)}`;
}

async function sendDiscordDebug(content) {
  if (!config.debugEnabled) return;
  try {
    await sendDiscord(`[bot-debug] ${String(content).slice(0, 1800)}`);
  } catch {
    // best effort debug only
  }
}

function debugAuthHeadersSnapshot() {
  return {
    ...authHeaders,
    Authorization: `Bot ${maskToken(config.token)}`
  };
}

function helpText() {
  return [
    "Commands:",
    "- `scan` → /api/scan-rss-feed",
    "- `cik <CIK>` → /api/cik-json?cik=<CIK>",
    "- `log` → /api/log",
    "- `help` → this message"
  ].join("\n");
}

async function discordRequest(path, init = {}) {
  const headers = { ...authHeaders, ...(init.headers || {}) };
  if (config.debugEnabled) {
    await sendDiscordDebug(`discordRequest path=${path}`);
    await sendDiscordDebug(
      `authHeaders=${JSON.stringify({ ...headers, Authorization: `Bot ${maskToken(config.token)}` })}`
    );
    await sendDiscordDebug(`token check -> ${maskToken(config.token)}`);
  }

  const res = await fetch(`${discordApi}${path}`, {
    ...init,
    headers
  });

  const text = await res.text();
  let body = null;
  try {
    body = text ? JSON.parse(text) : null;
  } catch {
    body = text || null;
  }

  if (!res.ok) {
    const errorText = `Discord API ${res.status}: ${text.slice(0, 500)}`;
    await sendDiscordDebug(errorText);
    throw new Error(errorText);
  }
  return body;
}

// ---- NEW: get bot id for state message ----
async function fetchBotId() {
  const me = await discordRequest("/users/@me");
  return me?.id || "<unknown>";
}
// -------------------------------------------

async function sendChannelMessage(channelId, content) {
  await discordRequest(`/channels/${channelId}/messages`, {
    method: "POST",
    body: JSON.stringify({ content: content.slice(0, 1900) })
  });
}

async function hitApi(path) {
  const res = await fetch(`${config.apiBaseUrl}${path}`);
  const text = await res.text();
  let body = text;
  try {
    body = JSON.parse(text);
  } catch {}
  if (!res.ok) {
    throw new Error(
      `API ${res.status}: ${(typeof body === "string" ? body : JSON.stringify(body)).slice(0, 700)}`
    );
  }
  return body;
}

function normalizeCik(value) {
  return value.replace(/\D/g, "").slice(0, 10);
}

async function loadChannelsIfNeeded() {
  if (trackedChannelIds.length > 0) return;
  const channels = await discordRequest(`/guilds/${config.guildId}/channels`);
  for (const channel of channels) {
    if (config.allowedChannelNames.includes(channel.name)) {
      trackedChannelIds.push(channel.id);
    }
  }
}

async function primeWatermarks() {
  for (const channelId of trackedChannelIds) {
    const messages = await discordRequest(`/channels/${channelId}/messages?limit=1`);
    if (messages[0]?.id) {
      lastSeenMessageId.set(channelId, messages[0].id);
    }
  }
}

function commandFromMessage(content) {
  const [command, ...rest] = content.trim().split(/\s+/);
  return { cmd: (command || "").toLowerCase(), args: rest };
}

async function processCommand(channelId, content, authorId) {
  const { cmd, args } = commandFromMessage(content);
  if (!["scan", "cik", "log", "help"].includes(cmd)) return;

  try {
    if (cmd === "help") {
      await sendChannelMessage(channelId, `<@${authorId}>\n${helpText()}`);
      return;
    }
    if (cmd === "scan") {
      await sendChannelMessage(channelId, `<@${authorId}> running scan-rss-feed...`);
      const data = await hitApi("/api/scan-rss-feed");
      const count = Array.isArray(data?.results) ? data.results.length : "?";
      await sendChannelMessage(channelId, `<@${authorId}> ✅ scan complete (${count} results)`);
      return;
    }
    if (cmd === "log") {
      await sendChannelMessage(channelId, `<@${authorId}> exporting logs...`);
      await hitApi("/api/log");
      await sendChannelMessage(channelId, `<@${authorId}> ✅ log export complete`);
      return;
    }
    if (cmd === "cik") {
      const cik = normalizeCik(args[0] || "");
      if (!cik) {
        await sendChannelMessage(channelId, `<@${authorId}> usage: \`cik <CIK>\``);
        return;
      }
      await sendChannelMessage(channelId, `<@${authorId}> fetching cik ${cik}...`);
      await hitApi(`/api/cik-json?cik=${cik}`);
      await sendChannelMessage(channelId, `<@${authorId}> ✅ cik-json complete for ${cik}`);
    }
  } catch (error) {
    const msg = error instanceof Error ? error.message : String(error);
    await sendChannelMessage(channelId, `<@${authorId}> ❌ command failed: ${msg.slice(0, 1500)}`);
  }
}

async function pollChannel(channelId) {
  const after = lastSeenMessageId.get(channelId);
  const query = after ? `?after=${after}&limit=50` : "?limit=25";
  const messages = await discordRequest(`/channels/${channelId}/messages${query}`);
  if (!Array.isArray(messages) || messages.length === 0) return;

  const ordered = messages
    .slice()
    .sort((a, b) => (BigInt(a.id) > BigInt(b.id) ? 1 : -1));

  for (const message of ordered) {
    lastSeenMessageId.set(channelId, message.id);
    if (message.author?.bot) continue;
    await processCommand(channelId, message.content || "", message.author.id);
  }
}

async function main() {
  if (config.debugEnabled) {
    await sendDiscordDebug(`debug mode ON; token check -> ${maskToken(config.token)}`);
    await sendDiscordDebug(`authHeaders snapshot -> ${JSON.stringify(debugAuthHeadersSnapshot())}`);
  }

  await loadChannelsIfNeeded();
  if (trackedChannelIds.length === 0) {
    throw new Error(
      "Could not locate #filings or #filings2. Set DISCORD_FILINGS_CHANNEL_ID / DISCORD_FILINGS2_CHANNEL_ID."
    );
  }

  await primeWatermarks();

  // ---- NEW: state message on start ----
  try {
    botId = await fetchBotId();
  } catch (e) {
    // best effort; don't block startup
    botId = "<unknown>";
    await sendDiscordDebug(`fetchBotId failed: ${e instanceof Error ? e.message : String(e)}`);
  }

  try {
    await sendDiscord(`Start polling... botId=${botId} freq=${config.pollMs}ms`);
  } catch (e) {
    // best effort; don't block startup
    await sendDiscordDebug(`startup state send failed: ${e instanceof Error ? e.message : String(e)}`);
  }
  // -----------------------------------

  console.log(`discord-control-bot polling every ${config.pollMs}ms`);
  console.log(`botId: ${botId}`);
  console.log(`Guild: ${config.guildId}; Channels: ${trackedChannelIds.join(", ")}`);

  setInterval(async () => {
    // prevent overlap if a tick runs longer than pollMs
    if (isTickRunning) return;
    isTickRunning = true;

    try {
      pollCount += 1;

      // ---- NEW: every 100 polls send state ----
      if (pollCount % 100 === 0) {
        try {
          await sendDiscord(`${pollCount} poll counts`);
        } catch {
          // best effort; do not crash polling
        }
      }
      // ----------------------------------------

      for (const channelId of trackedChannelIds) {
        try {
          await pollChannel(channelId);
        } catch (error) {
          console.error(`poll failed for ${channelId}:`, error);
        }
      }
    } finally {
      isTickRunning = false;
    }
  }, config.pollMs);
}

main().catch(async (error) => {
  const errMsg = error instanceof Error ? error.message : String(error);
  await sendDiscordDebug(`discord-control-bot failed to start: ${errMsg}`);
  console.error("discord-control-bot failed to start:", error);
  process.exit(1);
});

File: scripts/discord-control-bot.py
#!/usr/bin/env python3
import json
import os
import re
import time
import urllib.parse
import urllib.request
import urllib.error
from pathlib import Path


def load_env_file(path: str = '.env.local') -> None:
    p = Path(path)
    if not p.exists():
        return
    for line in p.read_text(encoding='utf-8').splitlines():
        line = line.strip()
        if not line or line.startswith('#') or '=' not in line:
            continue
        key, value = line.split('=', 1)
        key = key.strip()
        value = value.strip().strip('"').strip("'")
        if key and key not in os.environ:
            os.environ[key] = value


load_env_file()

CONFIG = {
    'token': os.getenv('DISCORD_BOT_TOKEN', ''),
    'guild_id': os.getenv('DISCORD_GUILD_ID', ''),
    'channel_ids': [
        os.getenv('DISCORD_FILINGS1_CHANNEL_ID', ''),
        os.getenv('DISCORD_FILINGS_CHANNEL_ID', ''),
        os.getenv('DISCORD_FILINGS2_CHANNEL_ID', ''),
    ],
    'api_base_url': os.getenv('BEACON_API_BASE_URL', 'http://127.0.0.1:3000'),
    'poll_ms': int(os.getenv('DISCORD_BOT_POLL_MS', '4000')),
    'openai_api_key': os.getenv('OPENAI_API_KEY', ''),
    'openai_model': os.getenv('OPENAI_MODEL', 'gpt-4o-mini'),
}
CONFIG['channel_ids'] = [c for c in CONFIG['channel_ids'] if c]

if not CONFIG['token']:
    raise RuntimeError('Missing DISCORD_BOT_TOKEN in .env.local')
if not CONFIG['guild_id']:
    raise RuntimeError('Missing DISCORD_GUILD_ID in .env.local')

DISCORD_API = 'https://discord.com/api/v10'
AUTH_HEADERS = {
    'Authorization': f"Bot {CONFIG['token']}",
    'Content-Type': 'application/json',
}

last_seen_by_channel: dict[str, str] = {}
conversation_state: dict[str, dict] = {}


def discord_request(path: str, method: str = 'GET', body: dict | None = None):
    data = None
    headers = dict(AUTH_HEADERS)
    if body is not None:
        data = json.dumps(body).encode('utf-8')
    req = urllib.request.Request(f"{DISCORD_API}{path}", method=method, headers=headers, data=data)
    try:
        with urllib.request.urlopen(req, timeout=20) as res:
            raw = res.read().decode('utf-8')
            return json.loads(raw) if raw else None
    except urllib.error.HTTPError as e:
        txt = e.read().decode('utf-8', errors='ignore')
        raise RuntimeError(f"Discord API {e.code}: {txt[:500]}") from e


def call_api(path: str):
    req = urllib.request.Request(f"{CONFIG['api_base_url']}{path}", method='GET')
    try:
        with urllib.request.urlopen(req, timeout=90) as res:
            raw = res.read().decode('utf-8')
            if 'application/json' in (res.headers.get('content-type') or ''):
                return json.loads(raw) if raw else {}
            return raw
    except urllib.error.HTTPError as e:
        txt = e.read().decode('utf-8', errors='ignore')
        raise RuntimeError(f"API {e.code}: {txt[:500]}") from e


def send_channel_message(channel_id: str, content: str):
    discord_request(f"/channels/{channel_id}/messages", method='POST', body={'content': content[:1900]})


def classify_intent_with_chatgpt(text: str):
    if not CONFIG['openai_api_key']:
        return ('none', None)

    prompt = {
        'model': CONFIG['openai_model'],
        'input': [
            {
                'role': 'system',
                'content': [
                    {
                        'type': 'input_text',
                        'text': (
                            'You classify Discord command intents for a SEC filing bot. '
                            'Return strict JSON only with keys: intent,arg. '
                            'Allowed intent: scan,cik,log,help,confirm_yes,confirm_no,none. '
                            'arg should be a 10-digit cik only when intent=cik; else null.'
                        )
                    }
                ]
            },
            {
                'role': 'user',
                'content': [
                    {'type': 'input_text', 'text': text}
                ]
            }
        ],
        'text': {
            'format': {
                'type': 'json_schema',
                'name': 'intent_schema',
                'schema': {
                    'type': 'object',
                    'properties': {
                        'intent': {
                            'type': 'string',
                            'enum': ['scan', 'cik', 'log', 'help', 'confirm_yes', 'confirm_no', 'none']
                        },
                        'arg': {
                            'anyOf': [
                                {'type': 'string', 'pattern': '^\\d{10}$'},
                                {'type': 'null'}
                            ]
                        }
                    },
                    'required': ['intent', 'arg'],
                    'additionalProperties': False
                }
            }
        }
    }

    req = urllib.request.Request(
        'https://api.openai.com/v1/responses',
        method='POST',
        headers={
            'Authorization': f"Bearer {CONFIG['openai_api_key']}",
            'Content-Type': 'application/json'
        },
        data=json.dumps(prompt).encode('utf-8')
    )
    try:
        with urllib.request.urlopen(req, timeout=20) as res:
            raw = res.read().decode('utf-8')
        data = json.loads(raw) if raw else {}
        text_out = data.get('output_text', '').strip()
        parsed = json.loads(text_out) if text_out else {}
        intent = parsed.get('intent', 'none')
        arg = parsed.get('arg')
        if intent == 'cik' and isinstance(arg, str):
            arg = arg.zfill(10)
        if intent not in {'scan', 'cik', 'log', 'help', 'confirm_yes', 'confirm_no', 'none'}:
            return ('none', None)
        return (intent, arg)
    except Exception:
        return ('none', None)


def intent_from_message(text: str):
    t = text.strip().lower()
    if not t:
        return ('none', None)

    cik_match = re.search(r'\bcik\s*(\d{1,10})\b', t)
    if cik_match:
        return ('cik', cik_match.group(1).zfill(10))

    if any(k in t for k in ['scan', 'rss', 'scan-rss-feed']):
        return ('scan', None)
    if any(k in t for k in ['log', 'export', 'state.json']):
        return ('log', None)
    if any(k in t for k in ['help', 'menu', 'what can you do']):
        return ('help', None)
    if t in {'yes', 'y', 'confirm', 'go', 'here we go'}:
        return ('confirm_yes', None)
    if t in {'no', 'n', 'cancel', 'stop'}:
        return ('confirm_no', None)

    # fallback to ChatGPT classifier for natural-language commands
    ai_intent, ai_arg = classify_intent_with_chatgpt(text)
    if ai_intent != 'none':
        return (ai_intent, ai_arg)

    return ('none', None)


def state_key(channel_id: str, author_id: str) -> str:
    return f"{channel_id}:{author_id}"


def ask_next(channel_id: str, author_id: str):
    send_channel_message(channel_id, f"<@{author_id}> What do you want to do next? You can say: scan rss, cik <CIK>, log, help.")


def process_intent(channel_id: str, author_id: str, text: str):
    key = state_key(channel_id, author_id)
    state = conversation_state.get(key, {'pending': None, 'arg': None})

    intent, arg = intent_from_message(text)

    if intent == 'help':
        send_channel_message(channel_id, f"<@{author_id}> I can help with: `scan rss`, `cik <CIK>`, `log`. I will confirm before running.")
        return

    if intent in {'scan', 'log', 'cik'}:
        state['pending'] = intent
        state['arg'] = arg
        conversation_state[key] = state
        if intent == 'scan':
            send_channel_message(channel_id, f"<@{author_id}> Next I will scan RSS feed. Please confirm by replying `yes`.")
        elif intent == 'log':
            send_channel_message(channel_id, f"<@{author_id}> Next I will export log/state. Please confirm by replying `yes`.")
        else:
            send_channel_message(channel_id, f"<@{author_id}> Next I will fetch cik-json for `{arg}`. Please confirm by replying `yes`.")
        return

    if intent == 'confirm_no':
        conversation_state[key] = {'pending': None, 'arg': None}
        send_channel_message(channel_id, f"<@{author_id}> Cancelled. Tell me your next command.")
        return

    if intent == 'confirm_yes' and state.get('pending'):
        pending = state['pending']
        pending_arg = state.get('arg')
        send_channel_message(channel_id, f"<@{author_id}> Here we go.")
        try:
            if pending == 'scan':
                data = call_api('/api/scan-rss-feed')
                count = len(data.get('results', [])) if isinstance(data, dict) else '?'
                send_channel_message(channel_id, f"<@{author_id}> Here is the result: scan completed with {count} result(s).")
            elif pending == 'log':
                call_api('/api/log')
                send_channel_message(channel_id, f"<@{author_id}> Here is the result: log export completed.")
            elif pending == 'cik' and pending_arg:
                call_api(f"/api/cik-json?cik={urllib.parse.quote(pending_arg)}")
                send_channel_message(channel_id, f"<@{author_id}> Here is the result: cik-json completed for {pending_arg}.")
            else:
                send_channel_message(channel_id, f"<@{author_id}> I don't have a valid pending action.")
        except Exception as e:
            send_channel_message(channel_id, f"<@{author_id}> Sorry, command failed: {str(e)[:1400]}")
        finally:
            conversation_state[key] = {'pending': None, 'arg': None}
            ask_next(channel_id, author_id)
        return

    if intent == 'none':
        if state.get('pending'):
            send_channel_message(channel_id, f"<@{author_id}> I am waiting for confirmation. Reply `yes` or `no`.")
        else:
            send_channel_message(channel_id, f"<@{author_id}> I can help. Say `scan rss`, `cik <CIK>`, `log`, or `help`.")


def prime_watermarks():
    for channel_id in CONFIG['channel_ids']:
        messages = discord_request(f"/channels/{channel_id}/messages?limit=1")
        if isinstance(messages, list) and messages and messages[0].get('id'):
            last_seen_by_channel[channel_id] = messages[0]['id']


def poll_channel(channel_id: str):
    after = last_seen_by_channel.get(channel_id)
    query = f"?after={after}&limit=50" if after else '?limit=25'
    messages = discord_request(f"/channels/{channel_id}/messages{query}")
    if not isinstance(messages, list) or not messages:
        return

    messages_sorted = sorted(messages, key=lambda m: int(m['id']))
    for m in messages_sorted:
        last_seen_by_channel[channel_id] = m['id']
        if m.get('author', {}).get('bot'):
            continue
        if m.get('guild_id') != CONFIG['guild_id']:
            continue
        process_intent(channel_id, m.get('author', {}).get('id', ''), m.get('content', ''))


def main():
    if not CONFIG['channel_ids']:
        raise RuntimeError('Missing DISCORD_FILINGS_CHANNEL_ID / DISCORD_FILINGS2_CHANNEL_ID')
    prime_watermarks()
    print(f"python discord bot ready; channels={','.join(CONFIG['channel_ids'])}; poll={CONFIG['poll_ms']}ms")

    while True:
        for channel_id in CONFIG['channel_ids']:
            try:
                poll_channel(channel_id)
            except Exception as e:
                print(f"poll failed for {channel_id}: {e}")
        time.sleep(max(CONFIG['poll_ms'], 500) / 1000.0)


if __name__ == '__main__':
    main()
